README.md
# Depotter

Depotter is intended to help you testing your financial strategies.
Depotter lets you create a plain txt file containing all the positions you've got (or would have invested in) and calculates the W/L and performance of each and all positions.

It uses the unoffical Yahoo Finance API.

## Showcase
![Showcase](resources/showcase.png)

## Build from source
Since the application is written in C++ with minimal library dependency you simply need to run
```sh
$ make
```
and the depot should be available to you as binary.

## Prerequisites
1. Install the [`cJSON`][cjson] library and add it to your library path.

## Usage
Create a file called `depot.txt` with following content.

### Sample Depot file
```
# Amount Name Yahoo-Symbol BuyPrice
8.3 Dax ^GDAXI 11003
1.03 MSCI-WLD EUNL.DE 45
```

*NOTE*: See the [`sample_depot_file.txt`](./sample_depot_file.txt) in this repository for a file with more than just two positions.

Then call
```sh
$ ./depot depot.txt
```

You can also use relative paths like
```sh
$ ./depot ../depot.txt
$ # or
$ ./depot ~/depot.txt
```


[cjson]: https://github.com/DaveGamble/cJSON#cmake


CHANGELOG.md
# Changelog

## Unreleased
### Refactor
- refactor: add src folder
- refactor: rename main.cpp to src/depotter.cpp
- refactor: tidy up CHANGELOG.md

## 0.2.0 - 2020-11-05
### Features
- use Yahoo finance api
- add [`sample_depot_file`](./sample_depot_file.txt)

## 0.1.0 - 2020-11-04
### Features
- basic depot calculcations and printDepot function


main.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include "stockUpdater.h"

using namespace std;

struct Position
{
    double amount;
    string name;
    double buyPrice;
    double currentValue;
};
typedef Position Position;

struct Depot
{
    vector<Position> positions;
};
typedef Depot Depot;

auto split(const std::string& value, char separator)
    -> std::vector<std::string>
{
    std::vector<std::string> result;
    std::string::size_type p = 0;
    std::string::size_type q;
    while ((q = value.find(separator, p)) != std::string::npos) {
        result.emplace_back(value, p, q - p);
        p = q + 1;
    }
    result.emplace_back(value, p);
    return result;
}

int longestName(vector<Position> positions)
{
    int minLength = 0;
    for (int i = 0; i < positions.size(); i++)
    {
        Position pos = positions[i];
        int length = pos.name.size();
        if (minLength <= length)
        {
            minLength = length + 1;
        }
    }
    return minLength;
}

string pad(string input, int length)
{
    int size = input.length();
    int missing = length - size;
    string missingString = "";
    for (int i = 0; i < missing; i++)
    {
        missingString = missingString + " ";
    }
    return input + missingString;
}

double getCurrentValue(Depot depot)
{
    double sum = 0;
    for (int i = 0; i < depot.positions.size(); i++)
    {
        Position pos = depot.positions[i];
        sum += pos.currentValue * pos.amount;
    }
    return sum;
}

double getBuyPrice(Depot depot)
{
    double sum = 0;
    for (int i = 0; i < depot.positions.size(); i++)
    {
        Position pos = depot.positions[i];
        sum += pos.buyPrice * pos.amount;
    }
    return sum;
}

Position newPosition(double amount, string name, double buyPrice, double currentValue)
{
    Position pos;
    pos.amount = amount;
    pos.name = name;
    pos.buyPrice = buyPrice;
    if (currentValue == -1)
    {
        currentValue = buyPrice;
    }
    pos.currentValue = currentValue;
    return pos;
}

double calculatePerformance(Depot depot)
{
    double buyValue = getBuyPrice(depot);
    double currentValue = getCurrentValue(depot);

    return (currentValue / buyValue) * 100 - 100;
}

void printDepot(Depot depot)
{
    cout << "Your Depot" << endl;
    cout << pad("Amount", 20) << " | " << pad("Name", 20) << " | " << pad("Buy Price", 15) << " | " << pad("Current Value", 15) << " | "
        << "Performance" << endl;
    for (int i = 0; i < depot.positions.size(); i++)
    {
        Position pos = depot.positions[i];
        cout
            << pad(to_string(pos.amount), 20) << " | "
            << pad(pos.name, 20) << " | "
            << pad(to_string(pos.buyPrice), 15) << " | "
            << pad(to_string(pos.currentValue), 15) << " | "
            << ((pos.currentValue / pos.buyPrice) - 1) * 100 << " %"
            << endl;
    }
    cout << endl
        << pad("Buy Price", 20) << " | " << to_string(getBuyPrice(depot)) << endl;
    cout << pad("Current Value", 20) << " | " << to_string(getCurrentValue(depot)) << endl;

    cout << pad("> Plain W/L", 20) << " | " << to_string(getCurrentValue(depot) - getBuyPrice(depot)) << endl;

    cout << endl
        << pad("Avg. Performance", 20) << " | " << calculatePerformance(depot) << " %" << endl;
}

Depot readDepotFromFile(string filename)
{
    Depot d;

    ifstream infile(filename);
    string content ((istreambuf_iterator<char>(infile)),
            (istreambuf_iterator<char>()));

    auto lines = split(content, '\n');
    for (string &line : lines) {
        if (line.rfind("#", 0) == 0) {
            continue;
        }
        vector<string> parts = split(line, ' ');
        if (parts.size() != 4) {
            continue;
        }
        double amount = stod(parts[0]);
        string name = parts[1];
        double buyPrice = stod(parts[2]);
        double currentValue = stod(parts[3]);
        d.positions.push_back(newPosition(amount, name, buyPrice, currentValue));
    }

    return d;
}

int main(int argc, char *argv[])
{
    char* filename = argv[1];
    if (filename == NULL) {
        cout << "Please provide a filename." << endl;
        return 127;
    }
    Depot d = readDepotFromFile(argv[1]);
    printDepot(d);
}


src/stockUpdater.cpp
#include <stdio.h>
#include <curl/curl.h>
#include <string>
#include <iostream>
#include <cjson/cJSON.h>

using namespace std;

size_t writeFunction(void *ptr, size_t size, size_t nmemb, std::string *data) {
    data->append((char*) ptr, size *nmemb);
    return size *nmemb;
}

char* constructAPIRequest(const char* symbol) {
    string symbol_string(symbol);
    char* urlPartOne = "https://query2.finance.yahoo.com/v10/finance/quoteSummary/"; 
    char *urlPartTwo = "?formatted=true&lang=de-DE&region=DE&modules=price";
    char *url;
    asprintf(&url, "%s%s%s", urlPartOne, symbol, urlPartTwo);

    return url;
}

double parseYahooResponse(const char* const response) {
    const cJSON *response_json = cJSON_Parse(response);
    const cJSON *quoteSummary = cJSON_GetObjectItemCaseSensitive(response_json, "quoteSummary");
    const cJSON *result = cJSON_GetObjectItemCaseSensitive(quoteSummary, "result");
    const cJSON *first_result = cJSON_GetArrayItem(result, 0);
    const cJSON *price = cJSON_GetObjectItemCaseSensitive(first_result, "price");
    const cJSON *regularMarketPrice = cJSON_GetObjectItemCaseSensitive(price, "regularMarketPrice");
    const cJSON *finalPrice = cJSON_GetObjectItemCaseSensitive(regularMarketPrice, "raw");
    return finalPrice->valuedouble;
}

double getStockPrice(const char* symbol) {
    auto curl = curl_easy_init();
    if (curl) {
        curl_easy_setopt(curl, CURLOPT_URL, constructAPIRequest(symbol));
        curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 1L);
        curl_easy_setopt(curl, CURLOPT_MAXREDIRS, 50L);
        curl_easy_setopt(curl, CURLOPT_TCP_KEEPALIVE, 1L);

        string response_string;
        string header_string;

        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeFunction);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_string);

        curl_easy_perform(curl);
        curl_easy_cleanup(curl);

        const char *response_chars = response_string.c_str();
        return parseYahooResponse(response_chars);
    }
    return -1;
}



src/depotter.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include "stockUpdater.h"

using namespace std;

struct Position
{
    double amount;
    string name;
    string symbol;
    double currentValue;
    double buyPrice;
};
typedef Position Position;

struct Depot
{
    vector<Position> positions;
};
typedef Depot Depot;

auto split(const std::string& value, char separator)
    -> std::vector<std::string>
{
    std::vector<std::string> result;
    std::string::size_type p = 0;
    std::string::size_type q;
    while ((q = value.find(separator, p)) != std::string::npos) {
        result.emplace_back(value, p, q - p);
        p = q + 1;
    }
    result.emplace_back(value, p);
    return result;
}

int longestName(vector<Position> positions)
{
    int minLength = 0;
    for (int i = 0; i < positions.size(); i++)
    {
        Position pos = positions[i];
        int length = pos.name.size();
        if (minLength <= length)
        {
            minLength = length + 1;
        }
    }
    return minLength;
}

string pad(string input, int length)
{
    int size = input.length();
    int missing = length - size;
    string missingString = "";
    for (int i = 0; i < missing; i++)
    {
        missingString = missingString + " ";
    }
    return input + missingString;
}

double getCurrentValue(Depot depot)
{
    double sum = 0;
    for (int i = 0; i < depot.positions.size(); i++)
    {
        Position pos = depot.positions[i];
        sum += pos.currentValue * pos.amount;
    }
    return sum;
}

double getBuyPrice(Depot depot)
{
    double sum = 0;
    for (int i = 0; i < depot.positions.size(); i++)
    {
        Position pos = depot.positions[i];
        sum += pos.buyPrice * pos.amount;
    }
    return sum;
}

Position newPosition(double amount, string name, string symbol, double buyPrice)
{
    Position pos;
    pos.amount = amount;
    pos.name = name;
    pos.symbol = symbol;
    pos.buyPrice = buyPrice;
    pos.currentValue = getStockPrice(symbol.c_str());
    return pos;
}

double calculatePerformance(Depot depot)
{
    double buyValue = getBuyPrice(depot);
    double currentValue = getCurrentValue(depot);

    return (currentValue / buyValue) * 100 - 100;
}

void printDepot(Depot depot)
{
    cout << "Your Depot" << endl;
    cout
        << pad("Amount", 20) << " | "
        << pad("Name", 20) << " | "
        << pad("Symbol", 10) << " | "
        << pad("Buy Price", 15) << " | "
        << pad("Current Value", 15) << " | "
        << "Performance" << endl;

    for (int i = 0; i < depot.positions.size(); i++)
    {
        Position pos = depot.positions[i];
        cout
            << pad(to_string(pos.amount), 20) << " | "
            << pad(pos.name, 20) << " | "
            << pad(pos.symbol, 10) << " | "
            << pad(to_string(pos.buyPrice), 15) << " | "
            << pad(to_string(pos.currentValue), 15) << " | "
            << ((pos.currentValue / pos.buyPrice) - 1) * 100 << " %"
            << endl;
    }
    cout << endl
        << pad("Buy Price", 20) << " | " << to_string(getBuyPrice(depot)) << endl;
    cout << pad("Current Value", 20) << " | " << to_string(getCurrentValue(depot)) << endl;

    cout << pad("> Plain W/L", 20) << " | " << to_string(getCurrentValue(depot) - getBuyPrice(depot)) << endl;

    cout << endl
        << pad("Avg. Performance", 20) << " | " << calculatePerformance(depot) << " %" << endl;
}

Depot readDepotFromFile(string filename)
{
    Depot d;

    ifstream infile(filename);
    string content ((istreambuf_iterator<char>(infile)),
            (istreambuf_iterator<char>()));

    auto lines = split(content, '\n');
    for (string &line : lines) {
        if (line.rfind("#", 0) == 0) {
            continue;
        }
        vector<string> parts = split(line, ' ');
        if (parts.size() != 4) {
            continue;
        }
        double amount = stod(parts[0]);
        string name = parts[1];
        string symbol = parts[2];
        double buyPrice = stod(parts[3]);
        d.positions.push_back(newPosition(amount, name, symbol, buyPrice));
    }

    return d;
}

int main(int argc, char *argv[])
{
    char* filename = argv[1];
    if (filename == NULL) {
        cout << "Please provide a filename." << endl;
        return 127;
    }
    Depot d = readDepotFromFile(argv[1]);
    printDepot(d);
}
